### PR TITLE
 added syncing of the token list based on balances

### PR DESCRIPTION
**Pull request by DmitryBespalov**
fyi added syncing of tokens from the server side

### CODE REVIEW

**Review by sche — APPROVED** [review]

File: MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionRelayService.swift [hunk 1]
```diff
@@ -80,6 +80,29 @@ public class HTTPGnosisTransactionRelayService: TransactionRelayDomainService {
         return true
     }
 
+    public func updateSafeInfo(safe: Address) {
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionRelayService.swift:83]: @sche This is the way to get safe info, maybe needs modifications for your purposes

File: MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionService.swift [hunk 2]
```diff
@@ -56,6 +55,26 @@ public class HTTPGnosisTransactionService: SafeTransactionDomainService {
         return result
     }
 
+    public func updateTokens(safe: Address) {
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionService.swift:58]: @sche this is the way to add tokens

File: MultisigWallet/MultisigWalletDomainModel/Wallet/Wallet.swift [hunk 3]
```diff
@@ -249,4 +249,43 @@ public class Wallet: IdentifiableEntity<WalletID> {
...
     }
 
+    public var hasWritePermission: Bool {
+        hasThisDeviceKeyAsOwner || hasAncestorWalletWithThisDeviceKeyAsOwner
+    }
+
+    private var hasThisDeviceKeyAsOwner: Bool {
+        if let device = owners.first(with: .thisDevice),
+            DomainRegistry.externallyOwnedAccountRepository.find(by: device.address) != nil {
+            return confirmationCount > 0
```
Comment (sche) [inline MultisigWallet/MultisigWalletDomainModel/Wallet/Wallet.swift:259]: this line is confusing for me. Why do you need to check confirmationCount?

File: MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionRelayService.swift (hunk 1 above)
Comment (sche) [reply MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionRelayService.swift:83]: 👍

File: MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionRelayService.swift [hunk 4]
```diff
@@ -80,6 +80,29 @@ public class HTTPGnosisTransactionRelayService: TransactionRelayDomainService {
...
+        do {
+            let info = try httpClient.execute(request: GetSafeRequest(safe: safe.value))
+            guard let wallet = DomainRegistry.walletRepository.find(address: safe), wallet.isReadyToUse else { return }
+            wallet.changeAddress(info.address.address)
```
Comment (sche) [inline MultisigWallet/MultisigWalletImplementations/Services/HTTPGnosisTransactionRelayService.swift:87]: redundant

**Review by sche — APPROVED** [review]

**Pull request merged**